src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -956,6 +1076,64 @@ public DocumentId AddTextContainer(SourceTextContainer textContainer, string ful
                 return documentId;
             }
 
+            public void AddDynamicFile(IDynamicFileInfoProvider fileInfoProvider, DynamicFileInfo fileInfo, ImmutableArray<string> folders)
+            {
+                var documentInfo = CreateDocumentInfoFromFileInfo(fileInfo, folders.IsDefault ? null : (IEnumerable<string>)folders);
+                var documentId = documentInfo.Id;
+
+                lock (_project._gate)
+                {
+                    var filePath = documentInfo.FilePath;
+                    if (_documentPathsToDocumentIds.ContainsKey(filePath))
+                    {
+                        throw new ArgumentException($"'{filePath}' has already been added to this project.", nameof(filePath));
+                    }
+
+                    _documentPathsToDocumentIds.Add(filePath, documentId);
+
+                    _documentIdToDynamicFileInfoProvider.Add(documentId, fileInfoProvider);
+
+                    if (_project._eventSubscriptionTracker.Add(fileInfoProvider))
+                    {
+                        // subscribe to the event when we use this provider the first time
+                        fileInfoProvider.Updated += _project.OnDynamicFileInfoUpdated;
+                    }
+
+                    if (_project._activeBatchScopes > 0)
+                    {
+                        _documentsAddedInBatch.Add(documentInfo);
+                    }
+                    else
+                    {
+                        // right now, assumption is dynamically generated file can never be opened in editor
+                        _project._workspace.ApplyChangeToWorkspace(w => _documentAddAction(w, documentInfo));
```

Comment:
i do not know what that comment means, or how it relates to the line of code below it.

Reply:
this is where dynamic file is first introudced roslyn workspace. so, I have put comment the assumption/constract we decide on this file here. if we have other place to put the comment in, let me know. 

same kind of comment seems there for other kind of thing like AddSourceTextContainer saying, all files added that way are assumed to be opened.

Reply:
i think teh problem is that the comment simply isn't clear in terms of what it means.